src/platforms/common/jtagtap.c

```diff
@@ -61,31 +67,47 @@ static void jtagtap_reset(void)
 #ifdef TRST_PORT
 	if (platform_hwversion() == 0) {
 		gpio_clear(TRST_PORT, TRST_PIN);
-		for (volatile size_t i = 0; i < 10000U; i++)
-			continue;
+		/* Hold low for approximately 1.5 ms */
+		platform_delay(1U); /* Requires SysTick interrupt to be unblocked */
 		gpio_set(TRST_PORT, TRST_PIN);
 	}
 #endif
 	jtagtap_soft_reset();
 }
 
+/* Busy-looping delay snippet for GPIO bitbanging (rely on inlining) */
+static inline void platform_delay_busy(const uint32_t loops)
+{
+#if 1
+	register uint32_t i = loops;
+	do {
+		__asm__("nop");
+	} while (--i > 0U);
+#else
+	for (register uint32_t counter = loops; counter > 0; --counter)
+		__asm__("nop");
+#endif
+}
+
 static bool jtagtap_next_clk_delay()
 {
 	gpio_set(TCK_PORT, TCK_PIN);
-	for (volatile uint32_t counter = target_clk_divider; counter > 0; --counter)
-		continue;
+	platform_delay_busy(target_clk_divider);
 	const uint16_t result = gpio_get(TDO_PORT, TDO_PIN);
 	gpio_clear(TCK_PORT, TCK_PIN);
-	for (volatile uint32_t counter = target_clk_divider; counter > 0; --counter)
-		continue;
+	platform_delay_busy(target_clk_divider);
 	return result != 0;
 }
 
+static bool jtagtap_next_no_delay() __attribute__((optimize(3)));
+
 static bool jtagtap_next_no_delay()
 {
 	gpio_set(TCK_PORT, TCK_PIN);
+	__asm__("nop" ::: "memory");
```

Comment:
Annotated for the readers. This routine is too fast in comparison to others, because it has nothing to do besides toggle the jTCK pin, and it certainly does not move data into (or out of) BMP. I may even want to add more nops or something similar to tone down this one.